File: Mac/ES_R_DevKit_Mac_Simple/MEME_Academic/ViewController.m
//
//  ViewController.m
//  MEME_Academic
//
//  Created by D-CLUE on 2017/03/22.
//  Copyright © 2017年 jins-jp. All rights reserved.
//

#import "ViewController.h"

@interface ViewController ()<NSComboBoxDelegate, NSComboBoxDataSource>

@end

@implementation ViewController
{
    MEMELib_Academic* memelib;
    boolean_t connected_flag;
    boolean_t measurement_flag;
}

// =============================================================================
#pragma mark - viewDidLoad
// =============================================================================
- (void)viewDidLoad {
    [super viewDidLoad];

    // Do any additional setup after loading the view.
    
    memelib = [[MEMELib_Academic alloc] init];
    memelib.delegate = self;

    connected_flag = false;
    measurement_flag = false;
    
    [self.button_StartScan setHidden:NO];
    [self.button_Connect setHidden:YES];
    [self.button_StartMeasurement setHidden:YES];
    
    [self.combobox_MEME removeAllItems];

    [self.combobox_SelectMode insertItemWithObjectValue:@"Full" atIndex:[self.combobox_SelectMode numberOfItems]];
    [self.combobox_SelectMode selectItemAtIndex:(0)];
    
    [self.combobox_TransSpeed insertItemWithObjectValue:@"100Hz" atIndex:[self.combobox_TransSpeed numberOfItems]];
    [self.combobox_TransSpeed selectItemAtIndex:(0)];
    
    [self.combobox_AccelRange insertItemWithObjectValue:@"±2G" atIndex:[self.combobox_AccelRange numberOfItems]];
    [self.combobox_AccelRange insertItemWithObjectValue:@"±4G" atIndex:[self.combobox_AccelRange numberOfItems]];
    [self.combobox_AccelRange insertItemWithObjectValue:@"±8G" atIndex:[self.combobox_AccelRange numberOfItems]];
    [self.combobox_AccelRange insertItemWithObjectValue:@"±16" atIndex:[self.combobox_AccelRange numberOfItems]];
    [self.combobox_AccelRange selectItemAtIndex:(0)];

    [self.combobox_GyroRange insertItemWithObjectValue:@"±250dps" atIndex:[self.combobox_GyroRange numberOfItems]];
    [self.combobox_GyroRange insertItemWithObjectValue:@"±500dps" atIndex:[self.combobox_GyroRange numberOfItems]];
    [self.combobox_GyroRange insertItemWithObjectValue:@"±1000dps" atIndex:[self.combobox_GyroRange numberOfItems]];
    [self.combobox_GyroRange insertItemWithObjectValue:@"±2000dps" atIndex:[self.combobox_GyroRange numberOfItems]];
    [self.combobox_GyroRange selectItemAtIndex:(0)];
}


// =============================================================================
#pragma mark - setRepresentedObject
// =============================================================================
- (void)setRepresentedObject:(id)representedObject {
    [super setRepresentedObject:representedObject];

    // Update the view, if already loaded.
}

// =============================================================================
#pragma mark - memePeripheralFoundDelegate
// =============================================================================
- (void)memePeripheralFoundDelegate:(uint32_t)result DeviceName:(NSString*)name uuid:(NSString*)uuid {
    if (result == MEMELIB_OK) {
        NSLog(@"memePeripheralFoundDelegate %d %@ %@",result, name, uuid);

        [self.combobox_MEME insertItemWithObjectValue:name atIndex:[self.combobox_MEME numberOfItems]];
        [self.combobox_MEME selectItemAtIndex:([self.combobox_MEME numberOfItems]-1)];

        [self.button_StartScan setHidden:NO];
        [self.button_Connect setHidden:NO];
        [self.button_StartMeasurement setHidden:YES];
    }
    else {
        NSLog(@"memePeripheralFoundDelegate %d",result);
        NSLog(@"Call : stopScanningPeripherals");
        [memelib stopScanningPeripherals];
    }
}

// =============================================================================
#pragma mark - memePeripheralConnectedDelegate
// =============================================================================
- (void)memePeripheralConnectedDelegate:(uint32_t)result {
    NSLog(@"memePeripheralConnectedDelegate : %d",result);
    connected_flag = true;
    self.button_Connect.title = @"Disconnect";
    self.label_StateConnect.stringValue = @"State : Connected";

    [self.button_StartScan setHidden:YES];
    [self.button_Connect setHidden:NO];
    [self.button_StartMeasurement setHidden:NO];
}

// =============================================================================
#pragma mark - memePeripheralDisconnectedDelegate
// =============================================================================
- (void)memePeripheralDisconnectedDelegate:(uint32_t)result {
    NSLog(@"memePeripheralDisconnectedDelegate : %d",result);
    connected_flag = false;
    self.button_Connect.title =@"Connect";
    self.label_StateConnect.stringValue = @"State : Disconnected";

    [self.button_StartScan setHidden:NO];
    [self.button_Connect setHidden:YES];
    [self.button_StartMeasurement setHidden:YES];
}

// =============================================================================
#pragma mark - memeAcademicFullDataReceivedDelegate
// =============================================================================
- (void)memeAcademicFullDataReceivedDelegate:(AcademicFullData *)data {
    //NSLog(@"memeAcademicFullDataReceivedDelegate");
    self.label_DataCnt.stringValue = [NSString stringWithFormat:@"%d", data.Cnt];
    self.label_DataAccX.stringValue = [NSString stringWithFormat:@"%d", data.AccX];
    self.label_DataAccY.stringValue = [NSString stringWithFormat:@"%d", data.AccY];
    self.label_DataAccZ.stringValue = [NSString stringWithFormat:@"%d", data.AccZ];
    self.label_DataGyroX.stringValue = [NSString stringWithFormat:@"%d", data.GyroX];
    self.label_DataGyroY.stringValue = [NSString stringWithFormat:@"%d", data.GyroY];
    self.label_DataGyroZ.stringValue = [NSString stringWithFormat:@"%d", data.GyroZ];
    self.label_DataEogL.stringValue = [NSString stringWithFormat:@"%d", data.EogL];
    self.label_DataEogR.stringValue = [NSString stringWithFormat:@"%d", data.EogR];
    self.label_DataEogH.stringValue = [NSString stringWithFormat:@"%d", data.EogH];
    self.label_DataEogV.stringValue = [NSString stringWithFormat:@"%d", data.EogV];
    self.label_DataBattLv.stringValue = [NSString stringWithFormat:@"%d", data.BattLv];
}

// =============================================================================
#pragma mark - button_StartScan_Tapped
// =============================================================================
- (IBAction)button_StartScan_Tapped:(id)sender {
    NSLog(@"Call : startScanningPeripherals");
    [self.combobox_MEME removeAllItems];
    [memelib startScanningPeripherals];
}

// =============================================================================
#pragma mark - button_Connect_Tapped
// =============================================================================
- (IBAction)button_Connect_Tapped:(id)sender {
    if (connected_flag == false) {
        NSLog(@"Call : connectPeripheral");
        NSString* name = [self.combobox_MEME stringValue];
        [memelib connectPeripheral:(NSString*)name];
    }
    else {
        NSLog(@"Call : disconnectPeripheral");
        //NSString* name = [self.combobox_MEME stringValue];
        [memelib disconnectPeripheral];
    }
}

// =============================================================================
#pragma mark - button_StartMeasurement_Tapped
// =============================================================================
- (IBAction)button_StartMeasurement_Tapped:(id)sender {
    NSInteger accelrange = [self.combobox_AccelRange indexOfSelectedItem];
    NSInteger gyrorange = [self.combobox_GyroRange indexOfSelectedItem];
    
    if (measurement_flag == false) {
        self.button_StartMeasurement.title = @"Stop Measurement";

        [self.button_StartScan setHidden:YES];
        [self.button_Connect setHidden:YES];
        [self.button_StartMeasurement setHidden:NO];

        [memelib setSelectMode:MEMEMode_Full];
        [memelib setTransMode:MEMEQuality_High];
        [memelib setAccelRange:(uint32_t)accelrange];
        [memelib setGyroRange:(uint32_t)gyrorange];
        measurement_flag = true;
        [memelib startDataReport];
    }
    else {
        self.button_StartMeasurement.title = @"Start Measurement";
        
        [self.button_StartScan setHidden:YES];
        [self.button_Connect setHidden:NO];
        [self.button_StartMeasurement setHidden:NO];

        measurement_flag = false;
        [memelib stopDataReport];
    }
}
@end

File: Mac/ES_R_DevKit_Mac_Simple/MEME_Academic/ViewController.swift
//
//  ViewController.swift
//  MEME_Academic
//
//  Created by JINS ASSIST開発実機検証  on 2026/06/02.
//  Copyright © 2026 jins-jp. All rights reserved.
//

import Cocoa

class ViewController: NSViewController {

    // MARK: - IBOutlets
    @IBOutlet weak var button_StartScan: NSButton!
    @IBOutlet weak var button_Connect: NSButton!
    @IBOutlet weak var combobox_MEME: NSComboBox!
    @IBOutlet weak var label_StateConnect: NSTextField!
    
    @IBOutlet weak var combobox_SelectMode: NSComboBox!
    @IBOutlet weak var combobox_TransSpeed: NSComboBox!
    @IBOutlet weak var combobox_AccelRange: NSComboBox!
    @IBOutlet weak var combobox_GyroRange: NSComboBox!
    
    @IBOutlet weak var button_StartMeasurement: NSButton!
    
    @IBOutlet weak var label_DataCnt: NSTextField!
    @IBOutlet weak var label_DataAccX: NSTextField!
    @IBOutlet weak var label_DataAccY: NSTextField!
    @IBOutlet weak var label_DataAccZ: NSTextField!
    @IBOutlet weak var label_DataGyroX: NSTextField!
    @IBOutlet weak var label_DataGyroY: NSTextField!
    @IBOutlet weak var label_DataGyroZ: NSTextField!
    @IBOutlet weak var label_DataEogL: NSTextField!
    @IBOutlet weak var label_DataEogR: NSTextField!
    @IBOutlet weak var label_DataEogH: NSTextField!
    @IBOutlet weak var label_DataEogV: NSTextField!
    @IBOutlet weak var label_DataBattLv: NSTextField!

    // MARK: - Private Properties
    private var memelib: (any MEMELibInterface)!
    private var connectedFlag = false
    private var measurementFlag = false

    // MARK: - View Life Cycle
    override func viewDidLoad() {
        super.viewDidLoad()

        memelib = MEMELibFactory.make()
        memelib.delegate = self

        connectedFlag = false
        measurementFlag = false
        
        button_StartScan.isHidden = false
        button_Connect.isHidden = true
        button_StartMeasurement.isHidden = true
        
        combobox_MEME.removeAllItems()

        combobox_SelectMode.addItem(withObjectValue: "Full")
        combobox_SelectMode.selectItem(at: 0)
        
        combobox_TransSpeed.addItem(withObjectValue: "100Hz")
        combobox_TransSpeed.selectItem(at: 0)
        
        combobox_AccelRange.addItems(withObjectValues: ["±2G", "±4G", "±8G", "±16G"])
        combobox_AccelRange.selectItem(at: 0)

        combobox_GyroRange.addItems(withObjectValues: ["±250dps", "±500dps", "±1000dps", "±2000dps"])
        combobox_GyroRange.selectItem(at: 0)
    }

    override var representedObject: Any? {
        didSet {
            // 必要に応じてビューを更新
        }
    }

    // MARK: - IBActions
    @IBAction func button_StartScan_Tapped(_ sender: Any) {
        print("Call : startScanningPeripherals")
        combobox_MEME.removeAllItems()
        _ = memelib.startScanningPeripherals()
    }

    @IBAction func button_Connect_Tapped(_ sender: Any) {
        if !connectedFlag {
            print("Call : connectPeripheral")
            let name = combobox_MEME.stringValue
            _ = memelib.connectPeripheral(deviceName: name)
        } else {
            print("Call : disconnectPeripheral")
            _ = memelib.disconnectPeripheral()
        }
    }

    @IBAction func button_StartMeasurement_Tapped(_ sender: Any) {
        let accelrange = combobox_AccelRange.indexOfSelectedItem
        let gyrorange = combobox_GyroRange.indexOfSelectedItem
        
        if !measurementFlag {
            button_StartMeasurement.title = "Stop Measurement"

            button_StartScan.isHidden = true
            button_Connect.isHidden = true
            button_StartMeasurement.isHidden = false

            _ = memelib.setSelectMode(MEMEMode_Full)
            _ = memelib.setTransMode(MEMEQuality_High)
            _ = memelib.setAccelRange(UInt32(accelrange))
            _ = memelib.setGyroRange(UInt32(gyrorange))
            
            measurementFlag = true
            _ = memelib.startDataReport()
        } else {
            button_StartMeasurement.title = "Start Measurement"
            
            button_StartScan.isHidden = true
            button_Connect.isHidden = false
            button_StartMeasurement.isHidden = false

            measurementFlag = false
            _ = memelib.stopDataReport()
        }
    }
}

// =============================================================================
// MARK: - MEMELibAcademicDelegate
// =============================================================================
extension ViewController: MEMELibAcademicDelegate {
    
    func memePeripheralFoundDelegate(result: UInt32, deviceName: String?, uuid: String?) {
        if result == MEMELIB_OK {
            print("memePeripheralFoundDelegate \(result) \(deviceName ?? "") \(uuid ?? "")")

            if let name = deviceName {
                combobox_MEME.addItem(withObjectValue: name)
                combobox_MEME.selectItem(at: combobox_MEME.numberOfItems - 1)
            }

            button_StartScan.isHidden = false
            button_Connect.isHidden = false
            button_StartMeasurement.isHidden = true
        } else {
            print("memePeripheralFoundDelegate \(result)")
            print("Call : stopScanningPeripherals")
            _ = memelib.stopScanningPeripherals()
        }
    }

    func memePeripheralConnectedDelegate(result: UInt32) {
        print("memePeripheralConnectedDelegate : \(result)")
        connectedFlag = true
        button_Connect.title = "Disconnect"
        label_StateConnect.stringValue = "State : Connected"

        button_StartScan.isHidden = true
        button_Connect.isHidden = false
        button_StartMeasurement.isHidden = false
    }

    func memePeripheralDisconnectedDelegate(result: UInt32) {
        print("memePeripheralDisconnectedDelegate : \(result)")
        connectedFlag = false
        button_Connect.title = "Connect"
        label_StateConnect.stringValue = "State : Disconnected"

        button_StartScan.isHidden = false
        button_Connect.isHidden = true
        button_StartMeasurement.isHidden = true
    }

    func memeAcademicFullDataReceivedDelegate(data: AcademicFullData) {
        label_DataCnt.stringValue = "\(data.cnt)"
        label_DataAccX.stringValue = "\(data.accX)"
        label_DataAccY.stringValue = "\(data.accY)"
        label_DataAccZ.stringValue = "\(data.accZ)"
        label_DataGyroX.stringValue = "\(data.gyroX)"
        label_DataGyroY.stringValue = "\(data.gyroY)"
        label_DataGyroZ.stringValue = "\(data.gyroZ)"
        label_DataEogL.stringValue = "\(data.eogL)"
        label_DataEogR.stringValue = "\(data.eogR)"
        label_DataEogH.stringValue = "\(data.eogH)"
        label_DataEogV.stringValue = "\(data.eogV)"
        label_DataBattLv.stringValue = "\(data.battLv)"
    }
}
